### PR TITLE
protocol/tx: return more specific types from newData and newSpend

### PR DESCRIPTION
**Pull request by bobg**
Also remove a stale TODO.

### CODE REVIEW

**Comment by bobg** [comment]
PTAL
I am smol

File: protocol/tx/data.go [hunk 1]
```diff
@@ -29,15 +29,13 @@ func (d *data) Body() interface{} { return d.body }
...
-func newData(hash bc.Hash) entry {
+func newData(hash bc.Hash) *data {
 	d := new(data)
 	d.body = hash
 	return d
 }
 
 func hashData(data []byte) (h bc.Hash) {
-	// TODO: Do we want domain separation here?  (E.g., a "data:"
-	// prefix.) If so, both the code and the spec need updating.
```
Comment (kr) [inline protocol/tx/data.go]: Did we make a call on this? IMO we should do domain separation as a matter of habit, unless there's a really clear reason not to. cc @tonychain

Comment (tonychain) [inline protocol/tx/data.go]: I'm definitely a fan of always doing domain separation. The overhead it adds is negligible and it's good security hygiene

Comment (jbowens) [inline protocol/tx/data.go]: @kr — I think this was the discussion around it:
https://chainhq.slack.com/archives/txgraph/p1486079399000187

**Comment by bobg** [comment]
PTAL

File: protocol/tx/data.go (hunk 1 above)
Comment (kr) [inline protocol/tx/data.go]: That the data entry itself serves as a domain separator seems plausible to me. I'll defer to @tonychain and @oleganza.

**Comment by jbowens** [comment]
lgtm

File: protocol/tx/data.go (hunk 1 above)
Comment (oleganza) [inline protocol/tx/data.go]: Data entry _is_ the domain separator for the raw data. This prevents having the raw hash of the data to be reused in the txgraph structure. For the purposes of identifying the raw data using `OP_REFDATAHASH` we need a raw hash that represents the raw data.

Comment (tonychain) [inline protocol/tx/data.go]: To back up a bit, the threat would be an attacker convincing you to sign a message in one context, and then reusing it in another.

So long as that threat is addressed, I'm happy. Sounds like it is?

Comment (oleganza) [inline protocol/tx/data.go]: Sounds like it is, yup. Thanks for clarifying the threat model, @tonychain!

**Pull request merged**